File: src/python/maps/map04.py
from browser import window
import ika
import sound
from thing import Thing

def AutoExec(engineRef):
    engineRef.mapThings.append(window.Snow.new(engineRef, 8000, [-.2, 3]))
    if 'waterrune' not in engineRef.saveFlags:
        engineRef.mapThings.append(RuneListener(engineRef))
    if 'nearend' in engineRef.saveFlags:
        engineRef.mapThings.append(RuneListener(engineRef))

def to3(engineRef):
    offset_from = 11 * 16  # first horizontal pos possible
    offset_to = 8 * 16  # first horizontal pos possible
    x = engineRef.player.sprite.x - offset_from + offset_to
    yield from engineRef.mapSwitchTask('map03.ika-map', (x, 1 * 16))

def to5(engineRef):
    yield from engineRef.mapSwitchTask('map05.ika-map', (10 * 16, 19 * 16))

class DeathListener(Thing):
    'Waits until the yeti is dead, then drops the fire rune.'
    def __init__(self, engineRef, yeti=None):
        self.engineRef = engineRef
        self.yeti = yeti

    def update(self):
        if self.yeti.stats.hp == 0:
            self.engineRef.playMusic("music/winter.ogg")
            self.engineRef.saveFlags['waterguard'] = 'True'
            return True

    def draw(self):
        pass

class RuneListener(object):
    def __init__(self, engineRef):
        self.engineRef = engineRef

    def update(self):
        if 'nearend' in self.engineRef.saveFlags and 'waterguard' not in self.engineRef.saveFlags:
            sound.playMusic("music/resurrection.it")
            y = window.yeti.SoulReaver.new(self.engineRef, self.engineRef.map.addSprite(15* 16, 17 * 16, self.engineRef.player.sprite.layer, 'soulreaver.ika-sprite'))
            self.engineRef.addEntity(y)
            self.engineRef.mapThings.append(DeathListener(self.engineRef, y))
            return True
        elif 'waterrune' in self.engineRef.saveFlags and 'nearend' not in self.engineRef.saveFlags:
            self.engineRef.addEntity(
                window.yeti.Yeti.new(self.engineRef, self.engineRef.map.addSprite(15* 16, 32 * 16, self.engineRef.player.sprite.layer, 'yeti.ika-sprite'))
            )
            return True

    def draw(self):
        pass

File: src/winter/maps/map04.ts
import { MapScript } from "./mapscript.js"
import { Thing } from "./../thing.js"
import { PyEngine, Snow } from "./../winter.js"
import { SoulReaver, Yeti } from "./../yeti.js"

export default new MapScript(autoexec, { to3, to5 })

export function autoexec(engineRef: PyEngine) {
    engineRef.addMapThing(new Snow(engineRef, 8000, [-.2, 3]))
    if (!engineRef.hasSaveFlag('waterrune')) {
        engineRef.addMapThing(new RuneListener(engineRef))
    }
    if (engineRef.hasSaveFlag('nearend')) {
        engineRef.addMapThing(new RuneListener(engineRef))
    }
}

export function *to3(engineRef: PyEngine) {
    const offset_from = 11 * 16  // first horizontal pos possible
    const offset_to = 8 * 16  // first horizontal pos possible
    const p = engineRef.getPlayerEntity()
    const x = p.sprite.x - offset_from + offset_to
    yield* engineRef.mapSwitchTask('map03.ika-map', [x, 1 * 16])
}

export function *to5(engineRef: PyEngine) {
    yield* engineRef.mapSwitchTask('map05.ika-map', [10 * 16, 19 * 16])
}

class DeathListener extends Thing {
    // Waits until the yeti is dead, then drops the fire rune.
    constructor(private engineRef: PyEngine, private yeti: Yeti) { super() }

    update() {
        if (this.yeti.stats.hp === 0) {
            this.engineRef.pyPlayMusic("music/winter.ogg")
            this.engineRef.setSaveFlag('waterguard', 'True')
            return true
        }
        return false
    }
}

class RuneListener extends Thing {
    constructor(private engineRef: PyEngine) { super() }

    update() {
        const engine = this.engineRef.getEngine()
        if (this.engineRef.hasSaveFlag('nearend')
            && !this.engineRef.hasSaveFlag('waterguard')
        ) {
            this.engineRef.pyPlayMusic("music/resurrection.it")
            const p = this.engineRef.getPlayerEntity()
            const y = new SoulReaver(this.engineRef, engine.map.addSprite(15* 16, 17 * 16, p.sprite.layer, 'soulreaver.ika-sprite'))
            this.engineRef.pyAddEntity(y)
            this.engineRef.addMapThing(new DeathListener(this.engineRef, y))
            return true
        } else if (this.engineRef.hasSaveFlag('waterrune') && !this.engineRef.hasSaveFlag('nearend')) {
            const p = this.engineRef.getPlayerEntity()
            this.engineRef.pyAddEntity(
                new Yeti(this.engineRef, engine.map.addSprite(15* 16, 32 * 16, p.sprite.layer, 'yeti.ika-sprite'))
            )
            return true
        }
        return false
    }
}
